pex/resolve/target_configuration.py

```diff
@@ -3,74 +3,130 @@
+class InterpreterConstraintsNotSatisfied(TargetConfigurationError):
+    """Indicates no interpreter meeting the requested constraints could be found."""
 
 
 @attr.s(frozen=True)
 class TargetConfiguration(object):
-    interpreters = attr.ib(
-        default=(), converter=_convert_interpreters
-    )  # type: Tuple[PythonInterpreter, ...]
-
-    platforms = attr.ib(
-        default=(), converter=convert_platforms
-    )  # type: Tuple[Optional[Platform], ...]
+    python_path = attr.ib(default=None)  # type: str
+    pythons = attr.ib(default=())  # type: Tuple[str, ...]
+    interpreter_constraints = attr.ib(default=())  # type: Tuple[Requirement, ...]
 
+    platforms = attr.ib(default=())  # type: Tuple[Optional[Platform], ...]
     assume_manylinux = attr.ib(default="manylinux2014")  # type: Optional[str]
-
-    @property
-    def interpreter(self):
-        # type: () -> Optional[PythonInterpreter]
-        if not self.interpreters:
-            return None
-        return PythonInterpreter.latest_release_of_min_compatible_version(self.interpreters)
-
-    def unique_targets(self):
-        # type: () -> OrderedSet[DistributionTarget]
-        def iter_targets():
-            # type: () -> Iterator[DistributionTarget]
-            if not self.interpreters and not self.platforms:
-                # No specified targets, so just build for the current interpreter (on the current
-                # platform).
-                yield DistributionTarget.current()
-                return
-
-            for interpreter in self.interpreters:
-                # Build for the specified local interpreters (on the current platform).
-                yield DistributionTarget.for_interpreter(interpreter)
-
-            for platform in self.platforms:
-                if platform is None and not self.interpreters:
-                    # Build for the current platform (None) only if not done already (ie: no
-                    # intepreters were specified).
-                    yield DistributionTarget.current()
-                elif platform is not None:
-                    # Build for specific platforms.
-                    yield DistributionTarget.for_platform(platform, manylinux=self.assume_manylinux)
-
-        return OrderedSet(iter_targets())
+    resolve_local_platforms = attr.ib(default=False)  # type: bool
+
+    def resolve_targets(self):
+        # type: () -> DistributionTargets
+        """Resolves the distribution targets satisfying the target configuration.
+
+        :raise: :class:`InterpreterNotFound` specific --python interpreters were requested but could
+            not be found.
+        :raise: :class:`InterpreterConstraintsNotSatisfied` if --interpreter-constraint were
+            specified but no conforming interpreters could be found.
+        """
+
+        # TODO(#1075): stop looking at PEX_PYTHON_PATH and solely consult the `--python-path` flag.
```

Comment:
This was moved from https://github.com/pantsbuild/pex/pull/1577/files#diff-6530c790096dac6af237266218e9b5934e25fee63241b7937ae66acc63581562L171